torch/csrc/jit/codegen/cuda/lower_thread_predicate.h

```diff
@@ -2,41 +2,47 @@
 #include <torch/csrc/WindowsTorchApiMacro.h>
 
 #include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
+#include <torch/csrc/jit/codegen/cuda/lower_utils.h>
 
 #include <bitset>
 
 namespace torch {
 namespace jit {
 namespace fuser {
 
-class TORCH_CUDA_API ThreadPredicates {
- private:
-  Fusion* fusion_;
+/*
+ * Map from tensorview to bit set represnting <BIDx, BIDy, BIDz, TIDx, TIDy,
+ * TIDz> If any dependency of TV had a parallelized reduction, we will track
+ * it here. This will be used for predicate generation to prevent
+ * parallelization on that axis. This is important if we have a reduction on
+ * for example TIDx, as the reduced value is only valid on threadIdx.x == 0
+ * therefore if we use that value later in the kernel we have that predicate.
+ * If we follow a reduction parallelized on TIDx with a broadcast on TIDx we
+ * no longer need the predicate and can reset the bit accordingly
+ */
+class TORCH_CUDA_API ThreadPredicateMap {
+ public:
+  using MapType =
+      std::unordered_map<const TensorView*, ir_utils::ParallelTypeBitmap>;
+  using const_iterator = MapType::const_iterator;
 
-  /*
-   * Map from tensorview to bit set represnting <BIDx, BIDy, BIDz, TIDx, TIDy,
-   * TIDz> If any dependency of TV had a parallelized reduction, we will track
-   * it here. This will be used for predicate generation to prevent
-   * parallelization on that axis. This is important if we have a reduction on
-   * for example TIDx, as the reduced value is only valid on threadIdx.x == 0
-   * therefore if we use that value later in the kernel we have that predicate.
-   * If we follow a reduction parallelized on TIDx with a broadcast on TIDx we
-   * no longer need the predicate and can reset the bit accordingly
-   */
-  std::unordered_map<const TensorView*, std::bitset<6>> thread_predicates;
+  ThreadPredicateMap(Fusion* _fusion);
 
-  // Update the thread_predicates bitset based on provided Expr
-  void updateBitSet(Expr*);
+  const_iterator find(const TensorView* tv) const;
+  const_iterator end() const;
+  const ir_utils::ParallelTypeBitmap& at(const TensorView* tv) const;
+  ir_utils::ParallelTypeBitmap& at(const TensorView* tv);
+  ir_utils::ParallelTypeBitmap& operator[](const TensorView* tv);
```

Comment:
no const overload for operator[] ?

Reply:
This is basically a wrapper around `std::unordered_map`, so I didn't choose to have the const overload.

Reply:
right, but at the same time we have overloads for at(), so it is inconsistent. If the const version of at() is not needed maybe we can remove that, but at and operator[] should be in sync.

Reply:
Yes, but the inconsistency is also the case for `std::unordered_map` as it has both const and non-const `at()`. The const version of `at()` is used at https://github.com/naoyam/pytorch/blob/block_broadcast/torch/csrc/jit/codegen/cuda/lower_utils.cpp#L591

Reply:
k, I see what you meant. thanks